hoot-core/src/main/cpp/hoot/core/ops/UnconnectedWaySnapper.cpp

```diff
@@ -579,6 +588,124 @@ bool UnconnectedWaySnapper::_snapUnconnectedWayEndNode(const NodePtr& unconnecte
+  {
+    WayPtr testWay = wayItr->second;
+    std::shared_ptr<geos::geom::Envelope> env2(testWay->getEnvelope(_map));
+
+    // If envelopes intersect, they will be close
+    if (LinearCriterion().isSatisfied(testWay) && env1->intersects(*env2))
+    {
+      // Now, do they cross?
+      if (ElementGeometryUtils::haveGeometricRelationship(wayToSnap, testWay, GeometricRelationship::Crosses, _map))
+      {
+        // Yes, now find intersection coords
+        // First, convert to geometries
+        ElementToGeometryConverter converter(_map);
+        std::shared_ptr<geos::geom::Geometry> lstr1 = converter.convertToGeometry(testWay);
+        std::shared_ptr<geos::geom::Geometry> lstr2 = converter.convertToGeometry(wayToSnap);
+
+        // Calculate intersections
+        std::unique_ptr<geos::geom::Geometry> pIntersection = lstr1->intersection(lstr2.get());
+
+        // Modify the map
+        std::unique_ptr<geos::geom::CoordinateSequence> seq = pIntersection->getCoordinates();
+        for (size_t i = 0; i < seq->getSize(); i++)
+        {
+          bool makeNewNode = true;
+
+          // Do we have node(s) here?
+          geos::geom::Envelope nodeEnv(seq->getAt(i));
+          nodeEnv.expandBy(_maxSnapDistance);
+          std::set<ElementId> neighborIds = SpatialIndexer::findNeighbors(nodeEnv, _crossingWayNodeIndex, _crossingWayNodeIndexToEid, _map, ElementType::Node);
+
+          if (1 == neighborIds.size())
+          {
+            long oldNodeId = neighborIds.begin()->getId();
+            NodePtr oldNode = _map->getNode(oldNodeId);
+            // Figure out what node the way belongs to, and add it to the OTHER one
+            if (wayToSnap->containsNodeId(oldNodeId))
+            {
+              long nodeIdx = WayUtils::closestWayNodeInsertIndex(oldNode, testWay, _map);
+              testWay->insertNode(nodeIdx, oldNodeId);
+              testWay->setStatus(Status::Conflated);
+              LOG_DEBUG(QString("Snapping crossing ways %1 and %2 with one old node %3").arg(wayToSnap->getId()).arg(testWay->getId()).arg(oldNodeId));
+            }
+            else if (testWay->containsNodeId(oldNodeId))
+            {
+              long nodeIdx = WayUtils::closestWayNodeInsertIndex(oldNode, wayToSnap, _map);
+              wayToSnap->insertNode(nodeIdx, oldNodeId);
+              wayToSnap->setStatus(Status::Conflated);
+              LOG_DEBUG(QString("Snapping crossing ways %1 and %2 with one old node %3").arg(wayToSnap->getId()).arg(testWay->getId()).arg(oldNodeId));
+            }
+            else
+            {
+              LOG_DEBUG(QString("Found node that doesn't belong to a way? %1").arg(oldNodeId));
+            }
+
+            makeNewNode = false;
+          }
+          else if (neighborIds.size() > 1)
+          {
+            // Uh, we have more than one node... find the one that is in the ref way (Unknown1) and keep it
+            for (auto it : neighborIds)
```

Comment:
Since `neighborIds` is a set of `ElementId` objects using a constant reference will reduce the number of copies made and speed up the iteration of the loop.
```suggestion
            for (const auto& it : neighborIds)
```